recipe/common.php

```diff
@@ -46,41 +47,142 @@
  * Configuration
  */
 
-set('keep_releases', 5);
+set('keep_releases', 5, function ($value) {
+    return is_integer($value) && $value >= -1;
+});
 
-set('repository', ''); // Repository to deploy.
+set('repository', '', function ($value) {
+    return is_string($value); 
+}); // Repository to deploy.
 
-set('shared_dirs', []);
-set('shared_files', []);
+set('shared_dirs', [], function ($value) {
+    if (!is_array($value)) {
+        return false;
+    }
+    foreach ($value as $dir) {
+        if (!is_dir($dir)) {
+            throw new ConfigurationException("Config error: folder \"$dir\" not found in \"shared_dirs\" option.");
+        }
+    }
+    return true;
+});
+
+set('shared_files', [], function ($value) {
+    if (!is_array($value)) {
+        return false;
+    }
+    foreach ($value as $file) {
+        if (!file_exists($file)) {
+            throw new ConfigurationException("Config error: file \"$file\" not found in \"shared_files\" option.");
+        }
+    }
+    return true;
+});
 
-set('copy_dirs', []);
+set('copy_dirs', [], function ($value) {
+    if (!is_array($value)) {
+        return false;
+    }
+    foreach ($value as $dir) {
+        if (!is_dir($dir)) {
+            throw new ConfigurationException("Config error: folder \"$dir\" not found in \"copy_dirs\" option.");
+        }
+    }
+    return true;
+});
 
-set('writable_dirs', []);
-set('writable_mode', 'acl'); // chmod, chown, chgrp or acl.
-set('writable_use_sudo', false); // Using sudo in writable commands?
-set('writable_recursive', true); // Common for all modes
-set('writable_chmod_mode', '0755'); // For chmod mode
-set('writable_chmod_recursive', true); // For chmod mode only (if is boolean, it has priority over `writable_recursive`)
+set('writable_dirs', [], function ($value) {
+    if (!is_array($value)) {
+        return false;
+    }
+    foreach ($value as $dir) {
+        if (!is_dir($dir)) {
+            throw new ConfigurationException("Config error: folder \"$dir\" not found in \"writable_dirs\" option.");
+        }
+    }
+    return true;
+});
 
-set('http_user', false);
-set('http_group', false);
+set('writable_mode', 'acl', function ($value) {
+    return $value === 'chmod'
+        || $value === 'chown'
+        || $value === 'chgrp'
+        || $value === 'acl'; 
+}); // chmod, chown, chgrp or acl.
 
-set('clear_paths', []);         // Relative path from release_path
-set('clear_use_sudo', false);    // Using sudo in clean commands?
+set('writable_use_sudo', false, function ($value) {
+    return is_bool($value);
```

Comment:
Maybe create some helpers for this? `isBool`

Reply:
Or just pass it as string `'is_bool'`.

Reply:
How about keeping the functions, but throwing specific errors as 'Config option "writable_use_sudo" is not a boolean'?

Or yet creating a helper that would receive 2 parameters, `$value` and a string to throw error?
```php
function isBool($value, $component) {
    if (!is_bool($value)) {
        throw new ConfigurationException("$component must have a boolean value");
    }
    return true;
}
```

Reply:
Yes, with throw it's okay. It will be cool to reduce duplicate code for bool case, as it really common.

Reply:
I came back on this helper function as it would include some not necessary complexity. I just used your `is_bool` string suggestion.